### PR TITLE
[WIP] Add http_parser.

### PR DESCRIPTION
**Pull request by norcalli**
Works just fine. Needs some bike shedding on the interface decisions and names
of the methods.

Currently adds:
- `vim.http_parse_url(url, [is_connect]) -> table`
- `vim.http_status_name(status_code) -> string`
- `vim._http_parser_create() -> parser` and `vim._http_parser_execute(parser, chunk) -> size_t`
- the `http_parser` metatable

Example usage:

```lua
local p = vim._http_parser_create()
repeat
  local bytes_parsed = vim._http_parser_execute(p, getchunk())
until p.complete

print(p.status_code, vim.http_status_name(p.status_code))
print(vim.inspect(p.headers))
print(p.body)
print(vim.inspect(p.table)) -- For everything.
```

- [ ] decide on `headers` structure. list of 2-tuples? dict of list? both in one table?
- [x] vim.http_parse_url or vim.parse_url?
- [ ] How to expose `vim._http_parser_create` and execute? Could use
a function wrap interface or coroutine.

```lua
function make_parser()
  local p = vim._http_parser_create()
  return function(chunk)
    if p.complete then
      return p.table
    end
    local bytes_parsed = vim._http_parser_execute(p, chunk)
    if p.complete then
      return p.table
    end
  end
end

local p = make_parser()
repeat until p(getchunk())
```
- [ ] move the code out of executor.c?
- [ ] add tests (make sure chunked transfer works)

### CODE REVIEW

File: src/nvim/lua/executor.c [hunk 1]
```diff
@@ -1008,3 +1012,244 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
...
+  FUNC_ATTR_NONNULL_ALL
+{
+  lua_pushliteral(lstate, "<http_parser>");
+  return 1;
+}
+
+static int nlua_http_parser_index(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+{
+  // TODO(ashkan): check this even for index?
+  // http_parser* p = luaL_checkudata(lstate, 1, "http_parser");
```
Comment (norcalli) [inline src/nvim/lua/executor.c]: This check is basically unnecessary if being used on __index but if some sick person decided to try to use the http_parser metatable elsewhere, the check would be useful.

File: src/nvim/CMakeLists.txt [hunk 2]
```diff
@@ -181,6 +183,8 @@ if(NOT MSVC)
 
   # tree-sitter: inlined external project, we don't maintain it. #10124
   set_source_files_properties(${TREESITTER_SOURCES} PROPERTIES COMPILE_FLAGS "${COMPILE_FLAGS} -Wno-conversion -Wno-pedantic -Wno-shadow -Wno-missing-prototypes -Wno-unused-variable")
+  # http-parser: inlined external project, we don't maintain it.
+  set_source_files_properties(${HTTP_PARSER_SOURCES} PROPERTIES COMPILE_FLAGS "${COMPILE_FLAGS} -Wno-conversion -Wno-pedantic -Wno-shadow -Wno-missing-prototypes -Wno-unused-variable")
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt]: Here and at the top of the module C source, we should give a URL (and perahaps a SHA if relevant). Same for treesitter.

File: src/nvim/lua/executor.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/lua/executor.c]: checks are good...

**Comment by norcalli** [comment]
Oh actually http headers should be a list instead of a dict since headers can be repeated.

**Comment by justinmk** [comment]
No objection to the general idea, though I wonder why inline? I see the code is here (and the readme is helpful):

https://github.com/nodejs/http-parser

Doesn't matter too much for now, we can focus on API for now. Revisit packaging later.

> move the code out of executor.c?

👍  The `#undef` stuff is a code smell IMO, and could be avoided then.

API suggestions:

- `vim.uri_parse` instead of `vim.http_parse_url`
- `new` is a better convention than `create`. Thus `vim.http_parser_new()`
- Instead of `p.table`, maybe `p.state` (or context, or result, ...)? "table" seems too generic.

File: src/nvim/lua/executor.c [hunk 3]
```diff
@@ -1008,3 +1012,253 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
...
+  size_t input_len = 0;
+  const char *input = luaL_checklstring(lstate, 1, &input_len);
+  int is_connect = (int)lua_toboolean(lstate, 2);
+
+  struct http_parser_url u;
+  http_parser_url_init(&u);
+  if (http_parser_parse_url(input, input_len, is_connect, &u) != 0) {
+    // We would give a more informative error, but http_parser doesn't give us
+    // any.
+    return luaL_error(lstate, "Failed to parse url.");
```
Comment (justinmk) [inline src/nvim/lua/executor.c]: we generally avoid punctuation in error messages where possible. So a period is inconsistent with most (should be _all_) other messages.

```suggestion
    return luaL_error(lstate, "Failed to parse url");
```

**Comment by norcalli** [comment]
@justinmk this isn't actually a `uri_parse` apparently, it's a `url_parse`. If you do `vim.uri_parse("file:///home/ashkan/")` it fails because it's expecting a host in that first segment. We could augment that if we wanted, but in the meantime that means I should rename it `url_parse`.

**Comment by teto** [comment]
we've decided against a http parser in another PR. Closing until we revisit that decision.

**Comment by bfredl** [comment]
A link to the conclusion? :) A specific lua rocks package that seems like a good choice?

**Comment by teto** [comment]
yes my bad https://github.com/neovim/neovim/pull/12459

**Pull request closed without merging**